lsp/gitlab_duo.lua

```diff
@@ -0,0 +1,440 @@
+    else
+      vim.notify('OAuth error: ' .. (body.error or 'unknown'), vim.log.levels.ERROR)
+    end
+  end
+
+  poll()
+end
+
+---@param bufnr integer
+---@param client vim.lsp.Client
+local function sign_in(_bufnr, client)
+  vim.notify('Starting GitLab device authorization...', vim.log.levels.INFO)
+
+  local auth_data = device_authorization()
+  if not auth_data then
+    return
+  end
+
+  vim.ui.open(auth_data.verification_uri .. '?user_code=' .. auth_data.user_code)
+
+  poll_for_token(auth_data.device_code, auth_data.interval or 5, client)
+end
+
+---@param client vim.lsp.Client
+local function sign_out(_, client)
+  local ok = os.remove(config.token_file)
+  if ok then
+    vim.notify('Signed out. Token removed.', vim.log.levels.INFO)
+    client.notify('workspace/didChangeConfiguration', {
+      settings = { token = '' },
+    })
+  else
+    vim.notify('Failed to remove token file', vim.log.levels.ERROR)
+  end
+end
+
+---@param client vim.lsp.Client
+local function show_status(_, _client)
+  local token_data = load_token()
+
+  if not token_data then
+    vim.notify('Not signed in. Run :LspGitLabDuoSignIn to authenticate.', vim.log.levels.INFO)
+    return
+  end
+
+  local info = {
+    'GitLab Duo Status:',
+    '',
+    'Instance: ' .. config.gitlab_url,
+    'Signed in: Yes',
+    'Has refresh token: ' .. (token_data.refresh_token and 'Yes' or 'No'),
+  }
+
+  if token_data.saved_at and token_data.expires_in then
+    local time_left = token_data.expires_in - (os.time() - token_data.saved_at)
+    if time_left > 0 then
+      local hours = math.floor(time_left / 3600)
+      local minutes = math.floor((time_left % 3600) / 60)
+      table.insert(info, string.format('Token expires in: %dh %dm', hours, minutes))
+    else
+      table.insert(info, 'Token status: EXPIRED')
+    end
+  end
+
+  vim.notify(table.concat(info, '\n'), vim.log.levels.INFO)
+end
+
+---@type vim.lsp.Config
+return {
+  cmd = {
+    'npx',
+    '--registry=https://gitlab.com/api/v4/packages/npm/',
+    '@gitlab-org/gitlab-lsp',
+    '--stdio',
+  },
+  root_markers = { '.git' },
+  filetypes = {
+    'ruby',
+    'go',
+    'javascript',
+    'typescript',
+    'typescriptreact',
+    'javascriptreact',
+    'rust',
+    'lua',
+    'python',
+    'java',
+    'cpp',
+    'c',
+    'php',
+    'cs',
+    'kotlin',
+    'swift',
+    'scala',
+    'vue',
+    'svelte',
+    'html',
+    'css',
+    'scss',
+    'json',
+    'yaml',
+  },
+  init_options = {
+    editorInfo = {
+      name = 'Neovim',
+      version = tostring(vim.version()),
+    },
+    editorPluginInfo = {
+      name = 'Neovim LSP',
+      version = tostring(vim.version()),
+    },
+    ide = {
+      name = 'Neovim',
+      version = tostring(vim.version()),
+      vendor = 'Neovim',
+    },
+    extension = {
+      name = 'Neovim LSP Client',
+      version = tostring(vim.version()),
+    },
+  },
+  settings = {
+    token = get_valid_token() or '',
```

Comment:
this only happens once, at module-load time. if the token expires what then? i think `settings` can be modified in an `on_init` handler.

Reply:
Then just set this to `''` ? It is misleading to just sprinkle a `get_valid_token()` call here, which again will have side effects like network calls at module-load time. It's better to do things in one place.

Reply:
@justinmk Thanks, I dropped method call from settings